### PR TITLE
GitPush step

### PR DESCRIPTION
**Pull request by p12tic**
This is an initial PR for GitPush step for pushing git commits to remote repository. The PR is WIP, this specific implementation has not yet been tested, only the unit tests pass. I've created the PR to gather early feedback.

The PR consists of two parts: first, the guts of Git step that are reusable have been extracted to util.git.GitWorkerCommonMixin (please advise of better name and/or location if possible); then the GitPush step has been implemented on top. The GitPush step is relatively simple for now, for example, it does not support the codebase parameter.

I've put the GitPush to steps.source.git module, but I think it's a little bit confusing because until now everything in steps.source is related to source checkout, not other operations. It probably makes sense to use the opportunity think about other non-checkout related steps, e.g. pushes on other VCs or commit operation.

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/steps/source/git.py [hunk 1]
```diff
@@ -28,11 +28,12 @@
 from buildbot.process import remotecommand
 from buildbot.process.properties import Properties
 from buildbot.steps.source.base import Source
-from buildbot.util.git import GitMixin
+from buildbot.steps.worker import CompositeStepMixin
+from buildbot.util.git import GitWorkerCommonMixin
```
Comment (tardyp) [inline master/buildbot/steps/source/git.py]: Common and Mixin looks like redounded. I would say GitStepMixin

File: master/buildbot/util/git.py [hunk 2]
```diff
@@ -87,3 +98,192 @@ def getSshWrapperScriptContents(keyPath, knownHostsPath=None):
 def getSshKnownHostsContents(hostKey):
     host_name = '*'
     return '{0} {1}'.format(host_name, hostKey)
+
+
+class GitWorkerCommonMixin(GitMixin):
+
+    def setupGitWorkerCommon(self):
```
Comment (tardyp) [inline master/buildbot/util/git.py]: setupGitStep

**Comment by tardyp** [comment]
image

**Comment by tardyp** [comment]
appart from this naming, the code looks good to me.

File: master/buildbot/util/git.py (hunk 2 above)
Comment (p12tic) [reply master/buildbot/util/git.py]: Fixed now.

File: master/buildbot/steps/source/git.py (hunk 1 above)
Comment (p12tic) [reply master/buildbot/steps/source/git.py]: Fixed now.

**Comment by p12tic** [comment]
The PR has been tested in production. The only issues observed were that sshPrivateKey does not work together with git submodules due to use of relative paths (the same issue affects regular Git step). I'll open a PR with a fix once this PR lands.

**Pull request merged**